dspy/clients/lm_local_arbor.py

```diff
@@ -0,0 +1,514 @@
+from typing import Any, Dict, List, Optional, Union, TypedDict, TYPE_CHECKING
+
+import openai
+
+import dspy
+from dspy.clients.provider import TrainingJob, ReinforceJob, Provider
+from dspy.clients.utils_finetune import TrainDataFormat, TrainingStatus, GRPOGroup, save_data
+
+
+if TYPE_CHECKING:
+    from dspy.clients.lm import LM
+
+
+class GRPOTrainKwargs(TypedDict):
+    num_generations: int
+
+
+class ArborTrainingJob(TrainingJob):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.provider_file_id = None
+        self.provider_job_id = None
+
+    def cancel(self):
+        if ArborProvider.does_job_exist(self.provider_job_id):
+            status = self.status()
+            if ArborProvider.is_terminal_training_status(status):
+                err_msg = "Jobs that are complete cannot be canceled."
+                err_msg += f" Job with ID {self.provider_job_id} is done."
+                raise Exception(err_msg)
+            openai.fine_tuning.jobs.cancel(self.provider_job_id)
+            self.provider_job_id = None
+
+        if self.provider_file_id is not None:
+            if ArborProvider.does_file_exist(self.provider_file_id):
+                openai.files.delete(self.provider_file_id)
+            self.provider_file_id = None
+
+        super().cancel()
+
+    def status(self) -> TrainingStatus:
+        status = ArborProvider.get_training_status(self.provider_job_id)
+        return status
+
+
+class ArborReinforceJob(ReinforceJob):
+    DEFAULT_TRAIN_KWARGS = {
+        "update_interval": 10,
+        "temperature": 0.9,
+        "beta": 0.04,
+        "num_iterations": 1,
+        "per_device_train_batch_size": 8,
+        "learning_rate": 1e-6,
+        "gradient_accumulation_steps": 1,
+        # This is false by default in TRL, but I think it makes sense to be true for us
+        "gradient_checkpointing": True, 
+        "lr_scheduler_type": "constant_with_warmup",
+        "max_prompt_length": None,
+        "max_completion_length": None,
+        "gradient_checkpointing_kwargs": None,
+        "bf16": False,
+        "scale_rewards": True,
+        "max_grad_norm": 1.0,
+        "report_to": "none",
+        "log_completions": True,
+        "logging_steps": 100,
+        # By default, none is the model's max context length
+        "max_context_length": None,
+        "lora": False
+    }
+
+    def __init__(self, lm: "LM", train_kwargs: GRPOTrainKwargs):
+        # The teleprompter must ensure that this is set
+        if "num_generations" not in train_kwargs:
+            raise ValueError("num_generations must be set in the training kwargs")
+
+        self.lm = lm
+        self.train_kwargs = train_kwargs
+        self.provider_job_id = None
+        self.checkpoints = {}
+        self.last_checkpoint = None
+
+    def initialize(self):
+        # TODO(GRPO Team): Set provider job ID
+        num_generations = self.train_kwargs.get("num_generations")
```

Comment:
I am wondering why every argument was explicitly extracted, and passed separately?